nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java

```diff
@@ -152,6 +160,9 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
       client.commitTable(base, metadata, newMetadataLocation, table, key);
       delete = false;
     } catch (NessieConflictException ex) {
+      if (ex instanceof NessieReferenceConflictException) {
+        handleReferenceConflictException((NessieReferenceConflictException) ex);
```

Comment:
threading through the code one wouldn't immediately expect `handleReferenceConflictException` to throw an exception. I think it would be better if `handleReferenceConflictException` would return `Optional<RuntimeException>`. 
Then the call in the `catch` clause could be `throw handleReferenceConflictException(ex).orElseGet(() -> new CommitFailedException(..))`.